### PR TITLE
Solved single value in data and breadcrumb range issues

### PR DESCRIPTION
**Pull request by victoria34**

### CODE REVIEW

File: app/_components/cbio-util.js [hunk 1]
```diff
@@ -666,6 +672,60 @@ cbio.util = (function() {
...
+    });
+
+    // Then sort
+    values.sort(function(a, b) {
+      return a - b;
+    });
+    
+    //make sure that min and max values are numbers.
+    for (head = 0; head < values.length; head++){
+      if ($.isNumeric(values[head])) {
```
Comment (zhx828) [inline app/_components/cbio-util.js]: No need to check  isNumeric any more. It has been down in line 688-689

File: app/_components/cbio-util.js [hunk 2]
```diff
@@ -666,6 +672,60 @@ cbio.util = (function() {
...
+        while (values[head] < 1) {
+          values[head] *= 10;
+          minZeros++;
+        }
+        break;
+      }
+    }
+    
+    for (tail = values.length - 1; tail >= 0; tail--) {
+      if ($.isNumeric(values[head])) {
```
Comment (zhx828) [inline app/_components/cbio-util.js]: Same here

File: app/scripts/views/components/barChart/barChart.js [hunk 3]
```diff
@@ -33,6 +33,22 @@
               break;
             }
           }
+        } else if (data_.smallDataFlag) {
+          if (d[data_.attrId] <= opts_.xDomain[1]) {
+            val = opts_.xDomain[0];
+          } else if (d[data_.attrId] > opts_.xDomain[opts_.xDomain.length - 3]) {
```
Comment (zhx828) [inline app/scripts/views/components/barChart/barChart.js]: Do we do checking on xDomain lenght?

**Review by zhx828 — CHANGES_REQUESTED** [review]
I need to talk about lines 163-258 in barChartTemplate.js

File: app/scripts/views/components/barChart/barChart.js (hunk 3 above)
Comment (victoria34) [reply app/scripts/views/components/barChart/barChart.js]: I think we check xDomain.length in function getTickFormat().

File: app/_components/cbio-util.js (hunk 1 above)
Comment (victoria34) [reply app/_components/cbio-util.js]: I will remove it.

File: app/_components/cbio-util.js (hunk 2 above)
Comment (victoria34) [reply app/_components/cbio-util.js]: I will remove it.

File: app/scripts/views/components/barChart/barChartTemplate.js [hunk 4]
```diff
@@ -192,31 +189,50 @@
 
       this.data.meta = _.map(_.filter(_.pluck(
         iViz.getGroupNdx(this.opts.groupid), this.opts.attrId), function(d) {
-        if (iViz.util.strIsNa(d, true)) {
+        if (typeof d === 'undefined' || d === 'na' || d === '' ||
```
Comment (zhx828) [inline app/scripts/views/components/barChart/barChartTemplate.js]: Use `iViz.util.strIsNa(d, true)` here

File: app/scripts/views/components/barChart/barChartTemplate.js [hunk 5]
```diff
@@ -229,13 +236,21 @@
           this.data.max = findExtremeResult[1];
         }
         
+        this.data.valueAsTick = false;
```
Comment (zhx828) [inline app/scripts/views/components/barChart/barChartTemplate.js]: Can we change this variable name? I don't understand it until I come here

File: app/scripts/views/components/barChart/barChart.js [hunk 6]
```diff
@@ -15,61 +15,86 @@
...
+                        if (d[data_.attrId] <= opts_.xDomain[i] &&
+                            d[data_.attrId] >= opts_.xDomain[i - 1]) {
+                            _min = opts_.xDomain[i - 1];
+                            _max = opts_.xDomain[i];
+                            val = _min + (_max - _min) / 3;
+                            break;
+                        }
+                    }
+                }
+            } else if (!data_.valueAsTick) {
```
Comment (zhx828) [inline app/scripts/views/components/barChart/barChart.js]: What happened if data.smaillDataFlag is true, and data_.valueAsTick is true as well?

**Review by zhx828 — CHANGES_REQUESTED** [review]

**Pull request merged**